exporters/otlp/lib/opentelemetry/exporters/otlp/exporter.rb

```diff
@@ -0,0 +1,252 @@
+            start_time_unix_nano: as_otlp_timestamp(span_data.start_timestamp),
+            end_time_unix_nano: as_otlp_timestamp(span_data.end_timestamp),
+            attributes: span_data.attributes&.map { |k, v| as_otlp_key_value(k, v) },
+            dropped_attributes_count: span_data.total_recorded_attributes - span_data.attributes&.size.to_i,
+            events: span_data.events&.map do |event|
+              Opentelemetry::Proto::Trace::V1::Span::Event.new(
+                time_unix_nano: as_otlp_timestamp(event.timestamp),
+                name: event.name,
+                attributes: event.attributes&.map { |k, v| as_otlp_key_value(k, v) },
+                # TODO: track dropped_attributes_count in Span#append_event
+              )
+            end,
+            dropped_events_count: span_data.total_recorded_events - span_data.events&.size.to_i,
+            links: span_data.links&.map do |link|
+              Opentelemetry::Proto::Trace::V1::Span::Link.new(
+                trace_id: link.context.trace_id,
+                span_id: link.context.span_id,
+                trace_state: link.context.tracestate,
+                attributes: link.attributes&.map { |k, v| as_otlp_key_value(k, v) },
+                # TODO: track dropped_attributes_count in Span#trim_links
+              )
+            end,
+            dropped_links_count: span_data.total_recorded_links - span_data.links&.size.to_i,
+            status: span_data.status&.yield_self do |status|
+              Opentelemetry::Proto::Trace::V1::Status.new(
+                code: status.canonical_code,
+                message: status.description,
+              )
+            end,
+          )
+        end
+
+        def as_otlp_timestamp(timestamp)
+          (timestamp.to_r * 1_000_000_000).to_i
+        end
+
+        def as_otlp_span_kind(kind)
+          case kind
+          when :internal then Opentelemetry::Proto::Trace::V1::Span::SpanKind::INTERNAL
+          when :server then Opentelemetry::Proto::Trace::V1::Span::SpanKind::SERVER
+          when :client then Opentelemetry::Proto::Trace::V1::Span::SpanKind::CLIENT
+          when :producer then Opentelemetry::Proto::Trace::V1::Span::SpanKind::PRODUCER
+          when :consumer then Opentelemetry::Proto::Trace::V1::Span::SpanKind::CONSUMER
+          else Opentelemetry::Proto::Trace::V1::Span::SpanKind::SPAN_KIND_UNSPECIFIED
+          end
+        end
+
+        def as_otlp_key_value(key, value)
+          kv = Opentelemetry::Proto::Common::V1::KeyValue.new(key: key)
+          kv.value = Opentelemetry::Proto::Common::V1::AnyValue.new
+          case value
+          when String
+            kv.value.string_value = value
+          when Integer
+            kv.value.int_value = value
+          when Float
+            kv.value.double_value = value
+          when true, false
+            kv.value.bool_value = value
+          # TODO: when Array
+          end
+          kv
```

Comment:
mostly just due to me not being very familiar with the `KeyValue` proto, but what would be the behavior for when a kv pair's value does not get matched by the case statement? does `kv.value` default to nil? and is that something that should be getting encoded and sent to the collector or should it get dropped pre-emptively, or is there a default that makes more sense?